pkg/sql/schemachanger/scbuild/builder.go

```diff
@@ -163,9 +164,40 @@ func (b *buildContext) build(ctx context.Context, n tree.Statement) (output scpb
 	default:
 		return nil, &notImplementedError{n: n}
 	}
+	// Optimize the generated nodes to apply simplification
+	// rules to avoid useless operations.
+	b.optimizeNodes(ctx)
```

Comment:
I strongly feel that this optimization phase should be distinct from `scbuild` and made into its own "phase", to borrow compiler terminology. Let's introduce a `scopt` package, with its own tests, etc. It's worth it, this way scbuild doesn't need to worry about being "efficient", only "correct".

Borrowing further from the compiler metaphor, it would make sense to decompose optimizations into distinct "passes". I realize that at this point in your PR there's effectively just one such "pass" but presumably there could be many more in time.